Separate wellcontributions for gpu

Tongdongq opened the pull request:
This PR adds the WellContributions class, and allows the cusparseSolver to be used without `--matrix-add-well-contributions=true. The wellcontributions can be processed coupled (like previous PR) or separate (default).

The use must pass
```
--use-gpu=true
```

Notes:
- Assumes `BlackoilWellModel` only has `StandardWell`, probably not scalable, but NORNE only uses these.
- All wells are applied simultaneously, this means that the columnIndices in `StandardWell::duneB_` and `duneC_` must be unique, to prevent write-after-write hazards.
- Applying the wellcontributions on GPU takes <1 second for a full NORNE run (< 0.7% of linear solve time).
- It might be better to do 1 `addMatrix()` call instead of 3 when adding the data of a well to the `WellContributions` object.
- A `WellContributions` object is created and filled for every linear solve. There is potential for reusing, but the number of wells and the size of their data can change per linear solve.
- The `COPY_ROW_BY_ROW` macro is defined in cusparseSolverBackend.cu, when true, it copies the nonzero values of the matrix row-by-row. Normally, these values are stored contiguously, but that is not guaranteed.

## Results ##
Intel Xeon E5-2620 v3 @ 2.4GHz, NVIDIA RTX GeForce 2080Ti

### Dune: ###
```
separate wellcontributions:
    Total runtime: 835 s
    Total linear solve: 480 s
    Newton Iterations: 1636
    Linear Iterations: 25110
```
```
coupled wellcontributions:
    Total runtime: 840 s
    Total linear solve: 477 s
    Newton Iterations: 1606
    Linear Iterations: 24078
```
### cusparse: ###
```
separate wellcontributions, this PR:
    Total runtime: 480 s
    Total linear solve: 138 s
    Newton Iterations: 1691
    Linear Iterations: 25608
```
```
coupled wellcontributions:
    Total runtime: 470 s
    Total linear solve: 137 s
    Newton Iterations: 1623
    Linear Iterations: 23956
```

The total runtime for coupled wellcontributions seems faster, but that is because the total assembly time varies a bit. The total linear solve time is roughly equal.

blattms commented:
I did some runs with NORNE on my not-so-shiny cpu wth the default setting:

| use_gpu | %mem | run-time | linear solve time |
|---|---|---|---|
| true | 0.7 | 545 | 223 |
| false | 0.4 | 630 | 303 |

For SPE9 the difference was negligible. For SPE1CASE2 the CPU was four times faster. Other models still need to be tested.

BTW using one core on my fancy multicore machine seems a bit unfair...

Then I got carried away and tried to combine MPI and GPU and that miserably, of course. The GPU code thinks it has the whole matrix but it hasn't and solves that unconditionally, the nonlinear solver is getting driven crazy, time steps get chopped, then your code tries to use the DUNE solvers on just a few processors and everyhing deadlocks.

That was to be expected and I think we need to check whether this is an mpi run with multiple processes and error out or deactivate the GPU in this case.

Review of the code will follow.

blattms commented:
For the larger model I run into a segmentation fault:
```
Report step  4/394 at day 35/6665, date = 05-Feb-2000

Time step 0, stepsize 12 days, at day 35/60, date = 05-Feb-2000
ESC[34;1m
Warning: NAN or INF value encountered during VFP calculation, the value is set to zeroESC[0m
ESC[34;1m
Warning: NAN or INF value encountered during VFP calculation, the value is set to zeroESC[0m
ESC[34;1m
Warning: NAN or INF value encountered during VFP calculation, the value is set to zeroESC[0m
ESC[34;1m
Warning: NAN or INF value encountered during VFP calculation, the value is set to zeroESC[0m
ESC[34;1m
Warning: NAN or INF value encountered during VFP calculation, the value is set to zeroESC[0m
ESC[34;1m
Warning: NAN or INF value encountered during VFP calculation, the value is set to zeroESC[0m
ESC[34;1m
Warning: NAN or INF value encountered during VFP calculation, the value is set to zeroESC[0m
ESC[34;1m
Warning: NAN or INF value encountered during VFP calculation, the value is set to zeroESC[0m
ESC[34;1m
Warning: NAN or INF value encountered during VFP calculation, the value is set to zeroESC[0m
ESC[34;1m
Warning: NAN or INF value encountered during VFP calculation, the value is set to zeroESC[0m
ESC[34;1m
Warning: Message limit reached for message tag: NAN_OR_INF_VFPESC[0m
[smaug:01959] *** Process received signal ***
[smaug:01959] Signal: Segmentation fault (11)
[smaug:01959] Signal code: Address not mapped (1)
[smaug:01959] Failing at address: 0x3e0
[smaug:01959] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x12730)[0x7fb6a8bbe730]
[smaug:01959] [ 1] /home/mblatt/src/dune/opm-2.6/opm-simulators/opm-parallel/bin/flow(_ZNK3Opm17BlackoilWellModelINS_10Properties4TTag14EclFlowProblemEE20getWellContributionsERNS_17WellContributionsE+0x120)[0x563658f754a0]
[smaug:01959] [ 2] /home/mblatt/src/dune/opm-2.6/opm-simulators/opm-parallel/bin/flow(_ZNK3Opm14ISTLSolverEbosINS_10Properties4TTag14EclFlowProblemEE31constructPreconditionerAndSolveILN4Dune14SolverCategory8CategoryE0ENS_22WellModelMatrixAdapterINS6_10BCRSMatrixINS_11MatrixBlockIdLi3ELi3EEESaISC_EEENS6_11BlockVectorINS6_11FieldVectorIdLi3EEESaISH_EEESJ_NS_17BlackoilWellModelIS3_EELb0EEENS6_3Amg21SequentialInformationEEEvRT0_RSJ_SR_RKT1_RNS6_21InverseOperatorResultE+0xb8f)[0x56365903f16f]
[smaug:01959] [ 3] /home/mblatt/src/dune/opm-2.6/opm-simulators/opm-parallel/bin/flow(_ZN3Opm14ISTLSolverEbosINS_10Properties4TTag14EclFlowProblemEE5solveERN4Dune11BlockVectorINS5_11FieldVectorIdLi3EEESaIS8_EEE+0xa42)[0x56365904ed32]
[smaug:01959] [ 4] /home/mblatt/src/dune/opm-2.6/opm-simulators/opm-parallel/bin/flow(_ZN3Opm17BlackoilModelEbosINS_10Properties4TTag14EclFlowProblemEE18nonlinearIterationINS_19NonlinearSolverEbosIS3_S4_EEEENS_15SimulatorReportEiRKNS_23SimulatorTimerInterfaceERT_+0xb6c)[0x56365904feac]
[smaug:01959] [ 5] /home/mblatt/src/dune/opm-2.6/opm-simulators/opm-parallel/bin/flow(_ZN3Opm24AdaptiveTimeSteppingEbosINS_10Properties4TTag14EclFlowProblemEE4stepINS_19NonlinearSolverEbosIS3_NS_17BlackoilModelEbosIS3_EEEEEENS_15SimulatorReportERKNS_14SimulatorTimerERT_bPKSt6vectorIiSaIiEE+0x7e9)[0x5636590523c9]
[smaug:01959] [ 6] /home/mblatt/src/dune/opm-2.6/opm-simulators/opm-parallel/bin/flow(_ZN3Opm34SimulatorFullyImplicitBlackoilEbosINS_10Properties4TTag14EclFlowProblemEE3runERNS_14SimulatorTimerE+0xe64)[0x563659055be4]
[smaug:01959] [ 7] /home/mblatt/src/dune/opm-2.6/opm-simulators/opm-parallel/bin/flow(_ZN3Opm12FlowMainEbosINS_10Properties4TTag14EclFlowProblemEE12runSimulatorEb+0x1a7)[0x563659057ff7]
[smaug:01959] [ 8] /home/mblatt/src/dune/opm-2.6/opm-simulators/opm-parallel/bin/flow(_ZN3Opm12FlowMainEbosINS_10Properties4TTag14EclFlowProblemEE7executeEiPPcbb+0x358)[0x5636590585b8]
[smaug:01959] [ 9] /home/mblatt/src/dune/opm-2.6/opm-simulators/opm-parallel/bin/flow(_ZN3Opm20flowEbosBlackoilMainEiPPcbb+0x11b)[0x563658e4c20b]
[smaug:01959] [10] /home/mblatt/src/dune/opm-2.6/opm-simulators/opm-parallel/bin/flow(main+0x1629)[0x563658cfc739]
[smaug:01959] [11] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xeb)[0x7fb69d87f09b]
[smaug:01959] [12] /home/mblatt/src/dune/opm-2.6/opm-simulators/opm-parallel/bin/flow(_start+0x2a)[0x563658d5e6aa]
[smaug:01959] *** End of error message ***
(END)
```
The warnings do also appear if I only use the CPU but it carries on. 

Looking into a debugger one sees that this is because that model must use multi segment wells:
```
(gdb) do
#2  0x00005555577fcfbb in Opm::BlackoilWellModel<Opm::Properties::TTag::EclFlowProblem>::getWellContributions (this=0x55555a70a640, wellContribs=...)
    at /home/mblatt/src/dune/opm-2.6/opm-simulators/opm/simulators/wells/BlackoilWellModel_impl.hpp:931
931	            derived->getWellSizes(nnz, numEq, numWellEq);
(gdb) p nnz
$3 = 21845
(gdb) l
926	    {
927	        for(unsigned int i = 0; i < well_container_.size(); i++){
928	            auto& well = well_container_[i];
929	            std::shared_ptr<StandardWell<TypeTag> > derived = std::dynamic_pointer_cast<StandardWell<TypeTag> >(well);
930	            unsigned int nnz, numEq, numWellEq;
931	            derived->getWellSizes(nnz, numEq, numWellEq);
932	            wellContribs.addSizes(nnz, numEq, numWellEq);
933	        }
934	        wellContribs.alloc_all();
935	        for(unsigned int i = 0; i < well_container_.size(); i++){
(gdb) p well
$4 = std::shared_ptr<Opm::WellInterface<Opm::Properties::TTag::EclFlowProblem>> (use count 1, weak count 0) = {get() = 0x55559e8f63f0}
(gdb) p derived
$5 = std::shared_ptr<Opm::StandardWell<Opm::Properties::TTag::EclFlowProblem>> (empty) = {get() = 0x0}
(gdb) 
```
The pointer cast failes and returns a nullptr.

We should check the result of the pointer cast and error out or recover otherwise at this point at the lastest. Hopefully we can do that earlier

Tongdongq commented:
I assume all Wells implement the WellInterface. We could add the addWellContribution() and getWellSizes() to that superclass, but I don't know how many types of wells need to be run on the GPU, and what their behaviour is. It looks like there is only StandardWell and MultisegmentWell.
BlackoilWellModel contains `well_container`, which is a `std::vector<std::shared_ptr<WellInterface<TypeTag>>>`. If only StandardWell has to be supported, perhaps a check could be done in BlackoilWellModel::createWellContainer(int).

I have also tried running the GPU with multiple mpi-processes in the past to no avail. This commit actually reaches Report step 43/247 before chopping.
I tried using isParallel() in ISTLSolverEbos(), but it did not work as expected. I put `if(gridForConn.comm().size() > 1 && use_gpu)` and that works.

Tongdongq commented on opm/simulators/linalg/ISTLSolverEbos.hpp:
```diff
@@ -240,11 +240,7 @@ class WellModelMatrixAdapter : public Dune::AssembledLinearOperator<M,X,Y>
             const double tolerance = EWOMS_GET_PARAM(TypeTag, double, LinearSolverReduction);
-            const bool matrix_add_well_contributions = EWOMS_GET_PARAM(TypeTag, bool, MatrixAddWellContributions);
             const int linear_solver_verbosity = parameters_.linear_solver_verbosity_;
-            if (use_gpu && !matrix_add_well_contributions) {
```
I put the check in the ISTLSolverEbos constructor, the GPU is disabled when MPI is used, but the warning print is a bit swamped by other warnings.

blattms replied:
Did you push the changes?

Tongdongq commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -121,7 +122,7 @@ void getSparsityPattern(BridgeMatrix& mat, std::vector<int> &h_rows, std::vector
 
 template <class BridgeMatrix, class BridgeVector>
-void BdaBridge::solve_system(BridgeMatrix *mat OPM_UNUSED, BridgeVector &b OPM_UNUSED, InverseOperatorResult &res OPM_UNUSED)
+void BdaBridge::solve_system(BridgeMatrix *mat OPM_UNUSED, BridgeVector &b OPM_UNUSED, WellContributions& wellContribs OPM_UNUSED, InverseOperatorResult &res OPM_UNUSED)
```
But if CUDA is not found by CMake, it is actually unused.

blattms replied:
It seems like it will never even be compiled in this case, see [here](/OPM/opm-simulators/blob/d067c08be7d91a0df380d96578a2d9487583891a/CMakeLists_files.cmake#L46-L48)

Tongdongq replied:
The BdaBridge is always compiled, but only created when CUDA is found. I think BdaBridge should also be in the CUDA-only list. And ISTLSolverEbos should only include the header if CUDA is found.
We could also add more #if HAVE_CUDA guards in the well files, since they don't need WellContributions.hpp if CUDA is not found.

blattms replied:
Nevermind. I missed that is compiled unconditionally.

Tongdongq commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -210,27 +211,31 @@ Dune::BCRSMatrix<Opm::MatrixBlock<double, 1, 1>, std::allocator<Opm::MatrixBlock
 Dune::BlockVector<Dune::FieldVector<double, 1>, std::allocator<Dune::FieldVector<double, 1> > > > \
 (Dune::BCRSMatrix<Opm::MatrixBlock<double, 1, 1>, std::allocator<Opm::MatrixBlock<double, 1, 1> > > *mat, \
     Dune::BlockVector<Dune::FieldVector<double, 1>, std::allocator<Dune::FieldVector<double, 1> > > &b, \
+    WellContributions& wellContribs, \
```
I think these are only required for preprocessor macros, I removed them.

Tongdongq commented on opm/simulators/linalg/bda/WellContributions.cu:
```diff
@@ -0,0 +1,420 @@
+                }
+            }
+            if(dbg)exit(0);
+#else
```
I had originally intended to perform `y -= (C*D*B) * x` instead of `y -= (C*(D*(B*x)))`, since that requires 1 spmv instead of 3.

blattms replied:
Do mind telling us what the result was and why you went the other way? Maybe we can learn something.

Tongdongq replied:
I first tried creating a new BCRSMatrix in ISTLSolverEbos, to contain all `(C*D*B)` of all wells. But the sparstity pattern of the wellcontributions changes often, and creating a new BCRSMatrix is not cheap. Luckily we found out that the `--matrix-add-wellcontributions` flag already prepares a matrix with all the indices it will ever need during the simulation.
I'm not sure why I didn't go for the first route for this PR, probably because the CPU does it like this. I started the implemention for `(C*D*B)`, but other things came up, and the second version worked and is fast enough (for NORNE).
The first version should use more memory, but it only uses square blocks, and I think it is a straightforward spmv. The second version is a bit more complex.

blattms commented:
Well curves match very good for SP1CASE2, SPE9, and [Norne](https://github.com/OPM/opm-simulators/files/4349400/gpu-wells-norne.pdf). For the latter there tiny differences due to different time stepping.

What I am wondering about, is whether our fallback to the CPU solvers still makes sense For the tests that we usually run our linear solver settings are quite heavily tuned. Usually the linear solver works pretty nicely and there is only time step reductions if the nonlinear solver does not converge. For these cases the fallback will nearly never be used. But if the time step reduction would occur because the linear solver does not converge the on gpu first the CPU linear solver would be tried and then time step would be chopped. Hence we would try to solve the linear system twice before reducing the time step. If this happens too often it might cost us some compute time.

Maybe in the future we want to think about skipping the fallback solve on the CPU.

Tongdongq commented:
The CPU stuff of WellContributions is only used in our CPU solver. We'll want an FPGA mode as well, but it is not needed in here.

blattms commented:
What is "our CPU solver" and when does it get used?

Tongdongq commented:
> When running a two phase problem (e.g. opm-tests/spe1/SPE1CASE2_OILGAS.DATA) the cuda solver often has convergence problems and we often fallback to the cpu. Are we maybe assuming three phases all the time?

Yes, at the time, we only had (or thought we had) three phase systems. The cusparse library supports any square blocks, so simply removing the check might solve the issue.
https://github.com/OPM/opm-simulators/blob/d067c08be7d91a0df380d96578a2d9487583891a/opm/simulators/linalg/bda/BdaBridge.cpp#L138-L141

> What is "our CPU solver" and when does it get used?

We made a CPU solver for debugging the GPU and FPGA outputs. It is not meant to be published.

Tongdongq commented on opm/simulators/linalg/bda/WellContributions.cu:
```diff
@@ -0,0 +1,420 @@
+        int idx = idx_b * blockDim.x + idx_t;
+        const unsigned int val_size = val_pointers[idx_b+1] - val_pointers[idx_b];
+
+        const int vals_per_block = dim * dim_wells;        // 12
```
No, it depends on numEq and numWellEq for StandardWell. Or at least the size of the block of StandardWell::duneB_. I'm not sure how the number of phases affects these values.

Tongdongq replied:
The two phase system with 2*3 blocks executes an invalid instruction in that kernel.

blattms replied:
Fixes for non three phase stuff should become their own PR.

Tongdongq commented on opm/simulators/linalg/bda/WellContributions.cu:
```diff
@@ -0,0 +1,420 @@
+
+            // merge all (dim) columns of 1 block, results in a single 1*dim_wells vector, which is used to multiply with invD
+            if(idx_t < vals_per_block){
+                // should be a loop as well, now only works for dim == 3
```
In the BdaBridge, but cusparse supports square blocks of arbitrary size.
This kernel for applying the wells really only works for blocks with 3 columns.

Tongdongq commented:
About the two phase warnings: the BdaBridge detects an unsupported blocksize, and disables itself with `use_gpu`. But ISTLSolverEbos does not know about this, and keeps sending new linear systems to BdaBridge. Since `result.converged` is not changed by the BdaBridge, ISTLSolverEbos assumes it failed to solve the system, but actually it did not even try to solve.

blattms commented:
Are you sure? Solving takes .7 seconds instead of .06 seconds.

Anyway thanks for the new commits. I will retry with the updated version. Maybe it is not an issue anymore.

BTW: Please fix these warnings:
```
/home/mblatt/src/dune/opm-2.6/opm-simulators/opm/simulators/linalg/bda/WellContributions.cu(53): warning: variable "idx" was declared but never referenced

/home/mblatt/src/dune/opm-2.6/opm-simulators/opm/simulators/linalg/bda/WellContributions.cu(62): warning: variable "NUM_THREADS" was declared but never referenced
```

blattms commented:
There seem to be some memory problems (which might have been there before). Please run opm-tests/spe1/SPE1CASE2_OILGAS with `--use-gpu=true` and see for yourself.

I also fired up a debugger. If I run the two-phase problem, then the bridge issues the waning, but does not return immediately. Instead it start allocating memory and seems to copy the sparsity pattern and it fires up a solve, nevertheless (maybe this causes the additional runtime?). For the subsequent requests to solve it does nothing but returning that it did not converge.

If I use `--use-gpu=false`, it still enters the solve_system method but just to return that it did not converge.

Tongdongq commented:
It should indeed return immediately. The first `cusparseSolver::solve_system()` also initializes the GPU, which takes some more time.
If `--use-gpu=false`, it should not print the warning that cusparseSolver did not converge. But perhaps I should rewrite that part a bit so that `bdaBridge->solve_system()` is not used?

blattms commented:
You misunderstood me, there is no warning. I mentioned the `--use-gpu=false`, because I would have expected the same behaviour when I use `--use-gpu=true``but for some reason we decided to treat as  `--use-gpu=false`. Currently, that is not the case, but it should be (at least it should not allocate and initialize any memory).

Tongdongq commented:
The SPE1CASE2_OILGAS case did trigger the unsupported warning and segfault at the end. I added a `return`, so preparations for the cusparseSolver are not done. I also added a guard in the cusparseSolver cleanup.
I still don't get your last point.

blattms commented:
If you have added a return in there now, you fixed my last point.

Thanks a lot. Will test.

blattms commented:
Cool the runtime is now approximately the same as one would expect.

Tongdongq commented on opm/simulators/linalg/bda/WellContributions.cu:
```diff
@@ -0,0 +1,420 @@
+            }
+            num_blocks += nnz;
+            dim = numEq;
+            dim_wells = numWellEq;
```
Done. I see MultisegmentWell has a similar `apply()`. I think `WellContributions` could support that as well.

blattms commented:
looks good now. I guess we should leave the rest (Macro, other phases, other wells) to other Prs.

blattms commented on opm/simulators/linalg/bda/cusparseSolverBackend.cu:
```diff
@@ -38,6 +38,10 @@
 
+// iff true, the nonzeroes of the matrix are copied row-by-row into a contiguous, pinned memory array, then a single GPU memcpy is done
+// otherwise, the nonzeroes of the matrix are assumed to be in a contiguous array, and a single GPU memcpy is enough
+#define COPY_ROW_BY_ROW 0
```
Leaving this for a future PR.

blattms commented:
jenkins build this please

blattms commented on CMakeLists.txt:
```diff
@@ -46,7 +46,8 @@ if(NOT CMAKE_DISABLE_FIND_PACKAGE_CUDA AND
   # Hence we call it unconditionally
-  find_package(CUDA)
+  # The WellContributions kernel uses __shfl_down_sync, which was introduced in CUDA 9.0
+  find_package(CUDA 9.0)
```
I do not think this is the correct way. It will only find CUDA compatible with version 9.0. That might be what we want (in the case that every higher version is compatible with the lower ones), but we cannot be sure. I guess we need to do:
```
find_package(CUDA);
if(CUDA_VERSION VERSION_LESS "9.0")
  set(CUDA_FOUND OFF)
endif()
```

blattms replied:
Merging anyway. Please open another PR.
